File: tree-hugger/lib/tests/fixtures/generics.py
"""Test module for Python generics."""
from typing import TypeVar, Generic, List, Optional, Callable

T = TypeVar('T')
U = TypeVar('U')

def identity(value: T) -> T:
    """Generic identity function."""
    return value

def map_list(items: List[T], func: Callable[[T], U]) -> List[U]:
    """Maps a list using a function."""
    return [func(item) for item in items]

def first_or_none(items: List[T]) -> Optional[T]:
    """Returns first item or None."""
    return items[0] if items else None

class Container(Generic[T]):
    """Generic container class."""
    
    def __init__(self, value: T) -> None:
        self._value = value
    
    def get(self) -> T:
        """Gets the contained value."""
        return self._value
    
    def map(self, func: Callable[[T], U]) -> 'Container[U]':
        """Maps the value using a function."""
        return Container(func(self._value))

File: tree-hugger/lib/tests/fixtures/generics.go
package main

// Identity returns the input value unchanged.
func Identity[T any](value T) T {
    return value
}

// Map applies a function to each element of a slice.
func Map[T, U any](items []T, fn func(T) U) []U {
    result := make([]U, len(items))
    for i, item := range items {
        result[i] = fn(item)
    }
    return result
}

// Container is a generic container type.
type Container[T any] struct {
    value T
}

// NewContainer creates a new container.
func NewContainer[T any](value T) *Container[T] {
    return &Container[T]{value: value}
}

// Get returns the contained value.
func (c *Container[T]) Get() T {
    return c.value
}

// Comparable is a constraint interface.
type Comparable interface {
    ~int | ~string
}

// Max returns the maximum of two comparable values.
func Max[T Comparable](a, b T) T {
    if a > b {
        return a
    }
    return b
}
